### ISSUE TITLE
nginx 1.9.9 not shutting down properly when on 1.10.33.2-beta

### ISSUE THREAD

**Issue by centminmod** [pagespeed/ngx_pagespeed#1082]
Just updated on CentOS 6.7, nginx 1.9.9 to 1.10.33.2-beta using the clang and devtoolset-3 gcc 4.9.1 workaround outlined at https://github.com/pagespeed/ngx_pagespeed/issues/1079 on my live forums running pagespeed at https://community.centminmod.com/

After update, and restarting nginx gives me the following error

```
Starting nginx: nginx: [emerg] bind() to 0.0.0.0:80 failed (98: Address already in use)
nginx: [emerg] bind() to 0.0.0.0:81 failed (98: Address already in use)
nginx: [emerg] bind() to 0.0.0.0:82 failed (98: Address already in use)
nginx: [emerg] still could not bind()
```

and ends up with failed nginx start, process list shows nginx processes are still running so stop didn't happen.

If I turn off ngx_pagespeed via pagespeed off;, nginx starts properly

**Comment by centminmod** [comment]
more investigation seems it is related to the 2nd method outlined at https://github.com/pagespeed/ngx_pagespeed/issues/1079 with export CFLAGS workaround as even downgrading to 1.10.33.1-beta didn't help nginx properly stop and start

1.10.33.1-beta worked with the initial workaround passing devtoolset-3 toolchain via exported CC variable.

**Comment by crowell** [comment]
so it works when built with gcc, but not clang?

**Comment by centminmod** [comment]
yeah with GCC 4.9.1 not clang

i see even with nginx stop command that processes are consuming cpu time

```
top -c
top - 19:27:00 up 112 days, 13:26,  1 user,  load average: 6.18, 4.12, 2.56
Tasks: 186 total,   4 running, 182 sleeping,   0 stopped,   0 zombie
Cpu0  : 99.7%us,  0.3%sy,  0.0%ni,  0.0%id,  0.0%wa,  0.0%hi,  0.0%si,  0.0%st
Cpu1  :100.0%us,  0.0%sy,  0.0%ni,  0.0%id,  0.0%wa,  0.0%hi,  0.0%si,  0.0%st
Cpu2  :100.0%us,  0.0%sy,  0.0%ni,  0.0%id,  0.0%wa,  0.0%hi,  0.0%si,  0.0%st
Cpu3  :100.0%us,  0.0%sy,  0.0%ni,  0.0%id,  0.0%wa,  0.0%hi,  0.0%si,  0.0%st
Mem:   4045132k total,  3384436k used,   660696k free,    73224k buffers
Swap:  4194300k total,   287800k used,  3906500k free,  1401772k cached

  PID USER      PR  NI  VIRT  RES  SHR S %CPU %MEM    TIME+  COMMAND                                                                                                                                                                                     
 5634 nginx     10 -10  707m  39m  12m R 139.2  1.0   4:39.67 nginx: worker process                                                                                                                                                                       
 5633 nginx     10 -10  707m  39m  12m R 132.6  1.0   4:21.58 nginx: worker process                                                                                                                                                                       
 5635 nginx     10 -10  707m  39m  12m R 126.6  1.0   4:25.87 nginx: worker process 
```

**Comment by centminmod** [comment]
with GCC 4.9.1 nginx starts and forums load but all css is broken https://community.centminmod.com/

**Comment by crowell** [comment]
what is broken with the CSS? on a cursory look it appears fine to me.

**Comment by centminmod** [comment]
Looks like it's working now with pagespeed off and pagespeed on for css files.. must of been some caching within the forum app.

To summarize so far

nginx 1.9.9 with ngx_pagespeed 1.10.33.1-beta and 1.10.33.2-beta

- with clang and any gcc-toolchain path devtoolset-3 workaround with GCC 4.9.1 results in nginx failing to properly stop and those Address already in use errors and nginx worker processes consuming large amounts of cpu resources
- with GCC compiled nginx via devtoolset-3, ngx_pagespeed 1.10 works fine and is currently in use and seems to be working on the forums

**Comment by arielesp** [comment]
I would just like to inform you that I am also having several problems with this last 1.10.XX version , nginx 1.9.x.

After the upgrade, the server keeps crashing .

At the moment , I have to use version 1.9.32.11

**Comment by jeffkaufman** [comment]
@arielesp Does it print anything in the error log when it crashes?

**Comment by arielesp** [comment]
So, dont crash stopping... but dont answer and need restart nginx

I already tested in 20 servers...

There are some bug in version 1.10... or with compiler ... I dont know...

This problem occurs with big sites ... 

server setup
32GB RAM
SSD 120GB
Intel Xeon E3 1245v2

I tested with nginx 1.9.9, only frontend / backend

And other with apache as backend

**Comment by jeffkaufman** [comment]
@arielesp It sounds to me like you're saying you're seeing 1.10 hang and be unresponsive, but it's not crashing, is that right?  Is there anything printed to the error logs when it goes unresponsive?

**Comment by arielesp** [comment]
@jeffkaufman I didnt checked. ... but I will reinstall again to check ...

**Comment by jeffkaufman** [comment]
@centminmod Over on #1114 you indicated that you were still having these problems with 1.10, and so are still on 1.9?

(I had thought this was working for you now.)

**Comment by centminmod** [comment]
@jeffkaufman yes i have it working in latest beta of my Centmin Mod LEMP stack if i use GCC devtoolset but I need for it to work with Clang too as for my stack the difference in compile speed and memory usage is enough to justify using Clang compiler and that doesn't work with ngx_pagespeed 1.10.

For now, I am just temporarily disabling ngx_pagespeed from integration into the stable version of my Centmin Mod LEMP stack (which doesn't have my fix from my beta version) as it's breaking a lot of users nginx 1.9.11 upgrades (~3,000 new installs per month). Guess this is expected with nginx mainline 1.9 usage :)

**Comment by crowell** [comment]
@centminmod sorry for the delay on this.

I've tried with building clang 3.7, with using devtoolset and with gcc5 compiled from source.

Each time it seems like the __init_array isn't populated when using clang. I'm opening a bug with clang to track this (maybe it's still our fault, but I'm not convinced otherwise).

**Comment by crowell** [comment]
bug i opened on llvm's bugtracker https://llvm.org/bugs/show_bug.cgi?id=26727

**Comment by centminmod** [comment]
@crowell cheers much appreciated - fingers crossed :)
